Add a JSON hostips output format

stephen-soltesz opened the pull request:
This change adds a new output format for the data in hostips and in JSON format.

This change is in support of a larger goal of publishing archival records of hostname & IP address assignments to GCS, under `gs://operator-<project>/metadata/v0/<date>/host-ips.json` (or similar).

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/m-lab/operator/217)
<!-- Reviewable:end -->

nkinkade commented:
Added a few comments.

---

Reviewed 1 of 3 files at r1.
Review status: 1 of 3 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---

*[plsync/mlabconfig.py, line 9 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BOpK27EjOmdzn8xGP:-L9BOpK27EjOmdzn8xGQ:b7plltp) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L9)):*
> ```Python
> import re
> import string
> import StringIO
> ```

Is the `StringIO` module used anywhere anylonger? Possibly same for the unittests.

---

*[plsync/mlabconfig.py, line 530 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BM3GA-DNtUp9cmbs4:-L9BM3GA-DNtUp9cmbs5:b-pn07yp) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L530)):*
> ```Python
> 
> def export_mlab_host_ips(sites, experiments):
>     """Writes csv data of all M-Lab servers and experiments to output."""
> ```

Does this docstring need to be updated ?

---

*[plsync/mlabconfig.py, line 790 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BNB3J3qcwOsfA6FXU:-L9BNB3J3qcwOsfA6FXV:b-6r7azx) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L790)):*
> ```Python
> 
>     if options.format == 'hostips':
>         output = export_mlab_host_ips(sites, experiments)
> ```

Here the function return is stored in `output`, below for `hostips-json` it is stroed to to `hostips`. Is it worth naming the variables consistently?

---

*[plsync/mlabconfig.py, line 791 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BMHJo2M-IYdXRK4Bt:-L9BMHJo2M-IYdXRK4Bu:b-en7bd4) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L791)):*
> ```Python
>     if options.format == 'hostips':
>         output = export_mlab_host_ips(sites, experiments)
>         #sys.stdout.write(output.getvalue())
> ```

Should this commented line just be removed?

---

*[plsync/mlabconfig.py, line 793 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BMSFs1gX512FYZpo5:-L9BMSFs1gX512FYZpo6:bx7q98v) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L793)):*
> ```Python
>         #sys.stdout.write(output.getvalue())
>         # Temporary workaround for HND01 load issues. Remove or generalize:
>         # https://github.com/m-lab/operator/issues/154
> ```

Should this comment be moved down, above the temporary TYO writes?

---

*[plsync/mlabconfig_test.py, line 118 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BOQeP47OLXg01vEmo:-L9BOQeP47OLXg01vEmp:bnm97ai) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig_test.py#L118)):*
> ```Python
>         for hostname, node in self.sites[0]['nodes'].iteritems():
>             experiments[0].add_node_address(node)
>         output = StringIO.StringIO()
> ```

Is the `StringIO.StringIO` object needed any longer?

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/m-lab/operator/217#-:-L9BOxBc0_U_qRjzFpo2:b-quu4xe)*
<!-- Sent from Reviewable.io -->

stephen-soltesz commented:
Review status: 1 of 4 files reviewed at latest revision, 6 unresolved discussions.

---

*[plsync/mlabconfig.py, line 9 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BOpK27EjOmdzn8xGP:-L9Batfg9_rSWMleH1st:b-ynyfwd) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L9)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

Is the `StringIO` module used anywhere anylonger? Possibly same for the unittests.
</blockquote></details>

Good catch. It looks like the .pylintrc config is far too permissive. So, I've also added a TODO to enable additional checks. I want to do that separately from this PR b/c that new config flags too many changes that are out of scope of this PR (but it's clearly needed).

---

*[plsync/mlabconfig.py, line 530 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BM3GA-DNtUp9cmbs4:-L9BceQQ0eJMn8sLqLFU:byvyago) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L530)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

Does this docstring need to be updated ?
</blockquote></details>

Yes. Done.

---

*[plsync/mlabconfig.py, line 790 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BNB3J3qcwOsfA6FXU:-L9BchSRDqZy6mAi-iYv:b2ba6mf) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L790)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

Here the function return is stored in `output`, below for `hostips-json` it is stroed to to `hostips`. Is it worth naming the variables consistently?
</blockquote></details>

Yes. And, actually, to prevent duplicating the static definitions for TYO, I've merged the hostips and hostips-json cases. I think this is better.

---

*[plsync/mlabconfig.py, line 791 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BMHJo2M-IYdXRK4Bt:-L9Bcq_mE_KE25zhKhEZ:b-896fix) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L791)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

Should this commented line just be removed?
</blockquote></details>

Done.

---

*[plsync/mlabconfig.py, line 793 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BMSFs1gX512FYZpo5:-L9BcrSuCOhfzk84cLf2:b-896fix) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig.py#L793)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

Should this comment be moved down, above the temporary TYO writes?
</blockquote></details>

Done.

---

*[plsync/mlabconfig_test.py, line 118 at r1](https://beta.reviewable.io/reviews/m-lab/operator/217#-L9BOQeP47OLXg01vEmo:-L9Bakzy7VYDsFxA0Iys:bss4a2b) ([raw file](https://github.com/m-lab/operator/blob/c82c0171bb51520758489c79b38f81d112e45458/plsync/mlabconfig_test.py#L118)):*
<details><summary><i>Previously, nkinkade wrote…</i></summary><blockquote>

Is the `StringIO.StringIO` object needed any longer?
</blockquote></details>

Removed.

---


*Comments from [Reviewable](https://beta.reviewable.io/reviews/m-lab/operator/217)*
<!-- Sent from Reviewable.io -->

nkinkade commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 3 of 3 files at r2.
Review status: 3 of 4 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://beta.reviewable.io/reviews/m-lab/operator/217#-:-L9Bfh6V5SeIi6SXhjJx:bnfp4nl)*
<!-- Sent from Reviewable.io -->
